doc/source/Code-Style-Guide.rst

```diff
@@ -0,0 +1,746 @@
+Code style guide
+================
+
+.. contents:: Contents
+    :depth: 3
+
+Developer advice
+----------------
+
+General comments
+~~~~~~~~~~~~~~~~
+
+-  IMPROVER uses a number of commonly used Python modules
+   e.g. `numpy <https://numpy.org/>`__,
+   `scipy <https://www.scipy.org/>`__. Additionally IMPROVER makes use
+   of `Iris <https://scitools.org.uk/iris/docs/latest/index.html>`__ for
+   the in-memory data representation.
+-  Aim for smaller methods with a single purpose.
+-  Aim for fewer logical branches in each method to make it easier to
+   test each route through the code and make it easier to maintain.
+-  Aim for simpler process methods (see the plugin example below),
+   preferably that just call other methods in the plugin.
+-  Centralise functions that check input (i.e. checking input is a cube)
+   so we only have to test them once.
+-  Aim for simple plugins with better defined purposes.
+-  Think about fixing metadata separately after main array manipulation
+   (although there are cases where it may make sense to do it together,
+   like when adding or removing dimensions).
+-  Some things that could be functions, should be functions, rather than
+   classes (if appropriate).
+-  Use inline comments where functionality is not immediately obvious,
+   eg where we do complex array or cube manipulation. Consider the top
+   level goal: that code should be readable to a non-expert developer.
+-  Document in plugin docstrings and CLI help when we add or remove a
+   dimension from a cube.
+-  Ensure that functions and classes are unit tested and CLIs (command
+   line interfaces) have acceptance tests.
+-  Use type annotations on function and method interfaces, except for
+   CLIs.
+
+Pull requests
+~~~~~~~~~~~~~
+
+Smaller pull requests (low hundreds of lines) to the IMPROVER code base
+are preferred to larger ones (thousand lines or higher), if reasonably
+possible. Very large pull requests are difficult to review and can lead
+to mistakes - but are sometimes necessary for more efficient or
+ambitious changes.
+
+In general, developers should consider whether a change that is becoming
+large (> 800 lines) can be submitted as a series of smaller,
+self-contained PRs for more manageable review. Reviewers should use
+their discretion in requesting PRs to be broken down if larger than
+this.
+
+Guidance (eg
+https://smartbear.com/learn/code-review/best-practices-for-peer-code-review/)
+suggests that for reviews to be effective reviewers should: \* Take time
+over reviews: don’t review faster than 500 lines of code per hour \* Not
+review more than 400 lines of code at a time, or for more than one hour
+continuously
```

Comment:
Another bulleted list that has been inlined.

Reply:
Fixed.